Move only disposer

p-podsiadly opened the pull request:
## Description

I came across this problem when writing an utility method, which registers the same comparator for multiple file extensions: [link](https://github.com/p-podsiadly/ImageApprovals/blob/43a363ae6f99994a76385a645820291e6afb1e65/ImageApprovals/src/Comparator.cpp#L41).

In linked code, `Disposer` class wraps a vector of `ComparatorDisposer` objects. Each `ComparatorDisposer` is created by `FileApprover::registerComparatorForExtension`, copied to a vector and destructed and at that point the comparator is immediately disposed.

Expected behavior is that the comparator is disposed once the copy in the vector is destructed. Destructor of the original `ComparatorDisposer` instance should not dispose the comparator.

## The solution

I've added a flag `isActive`, which is initially set to `true`. The flag is set to `false` when a `ComparatorDisposer` is moved-from, to indicate that it's destructor should not restore `previousComparator`.

Additionally, I've declared copy constructor as deleted.

claremacrae commented:
Hello, thank you very much for this, @p-podsiadly - I looked at the linked code and it was very exciting to see an image comparison implementation, especially one that doesn’t depend on Qt.

Would you be able to write a test for the change, that shows how it was broken, and the new behaviour, please?

I’d be worried otherwise that we might accidentally break it in future.

p-podsiadly commented:
Hello, @claremacrae, I've added the test - fails, as expected, without the fix. 

I'm not sure if the commit message should start with "F!!" or "B!!" according to the convention you're using - should I change it?

claremacrae commented:
Hello @p-podsiadly Can you spare a few minutes to pair on this - it'll be a lot more quicker than going back and forwards here, I feel...

claremacrae commented:
> Hello @p-podsiadly Can you spare a few minutes to pair on this - it'll be a lot more quicker than going back and forwards here, I feel...

I was struggling to work out how the test works, but I think I've got it now - so I'll accept the PR then add some comments to the test, and ask you to confirm if I've understood it.

Thank you very much indeed for doing this - I think I now need to work out how to effectively apply similar changes to all the other disposers.

claremacrae commented:
> I'm not sure if the commit message should start with "F!!" or "B!!" according to the convention you're using - should I change it?

I'm happy for it to stay as F!! and count it as a feature!

p-podsiadly commented:
> I was struggling to work out how the test works, but I think I've got it now - so I'll accept the PR then add some comments to the test, and ask you to confirm if I've understood it.

If you'd like to, I'll be able to talk later today.

I guess that the `"foo"`/`"bar"` comparisons in `TestComparator` could be confusing. Maybe it will be better if I changed this to case-insensitive comparison and rename `TestComparator` to `CaseInsensitiveComparator`?

PS I've added another test case to check if `ComparatorDisposer` is *not* copyable.

p-podsiadly commented:
Sorry, I've clicked "Close and comment" instead of "Comment" by mistake.

claremacrae commented:
> Sorry, I've clicked "Close and comment" instead of "Comment" by mistake.

Been there, done that - it's easily done! 😀 

> If you'd like to, I'll be able to talk later today. 

OK, thanks. For now I'll pop a few comments in code review, to get your thoughts on them, if you have time...

claremacrae commented on tests/Catch2_Tests/comparators/ComparatorDisposerTest.cpp:
```diff
@@ -0,0 +1,56 @@
+        disposer.reset(new ComparatorDisposer(std::move(temporaryDisposer)));
+    }
+
+    REQUIRE_NOTHROW(Approvals::verify("foo"));
```
If this fails, users see a diffing tool, and this output:

```
/Users/clare/Documents/develop/ApprovalTests/ApprovalTests.cpp/tests/Catch2_Tests/comparators/ComparatorDisposerTest.cpp:51: FAILED:
  REQUIRE_NOTHROW( Approvals::verify("foo") )
due to unexpected exception with message:
  Failed Approval: 
  Received does not match approved 
  Received : "/Users/clare/Documents/develop/ApprovalTests/ApprovalTests.cpp/
  tests/Catch2_Tests/comparators/approval_tests/ComparatorDisposerTest.
  ComparatorDisposer_object_can_be_moved_without_disposing_a_comparator_twice.
  received.txt" 
  Approved : "/Users/clare/Documents/develop/ApprovalTests/ApprovalTests.cpp/
  tests/Catch2_Tests/comparators/approval_tests/ComparatorDisposerTest.
  ComparatorDisposer_object_can_be_moved_without_disposing_a_comparator_twice.
  approved.txt"
```

In that situation, someone who didn't understand this test might be inclined to approve the new output, at which point the test would always pass, whilst not actually testing the intended behaviour...

One way to detect and prevent this would be to repeat the `REQUIRE_NOTHROW(Approvals::verify("foo"));`, both before and after the `disposer.reset` line...

p-podsiadly replied:
I changed the test to use `FileApprover::verify` instead of `Approvals::verify` so that the approved file does not need to be put in git and diff tool is not run.

claremacrae commented on tests/Catch2_Tests/comparators/approval_tests/ComparatorDisposerTest.ComparatorDisposer_object_can_be_moved_without_disposing_a_comparator_twice.approved.txt:
```diff
@@ -0,0 +1 @@
+bar
```
If this test continues to use an output file, could its file name be shortened by at least this many characters please: `comparator_twice.approved.txt`

**Background**: I've previously had problems cloning this reopen Windows, because of long file names of approval files exceeding the path length limit - and so having to rename the cloned folder to shorten the filename.

Because of this, I recently shortened the names of the longest files here.

Before this commit, `git ls-files` tells me the longest filename was:

`tests/DocTest_Tests/docs/approval_tests/DocTestDocsSamples.ApprovalTests-MultipleOutputFiles-AutoApproving.American.approved.txt`

This name is a fair bit longer:

`tests/Catch2_Tests/comparators/approval_tests/ComparatorDisposerTest.ComparatorDisposer_object_can_be_moved_without_disposing_a_comparator_twice.approved.txt`

p-podsiadly replied:
Now the "approved" file is written at runtime.

claremacrae commented on tests/Catch2_Tests/comparators/ComparatorDisposerTest.cpp:
```diff
@@ -0,0 +1,56 @@
+        auto temporaryDisposer = FileApprover::registerComparatorForExtension(
+            ".txt", std::make_shared<TestComparator>());
+
+        disposer.reset(new ComparatorDisposer(std::move(temporaryDisposer)));
```
It's really helpful to see how you guaranteed that a move operation would happen, for the testing. 

An explanatory comment would be really helpful, to explain the purpose of lines 46, 49 and 52.

p-podsiadly replied:
I've added comments before each step in the test.

claremacrae commented on tests/Catch2_Tests/comparators/ComparatorDisposerTest.cpp:
```diff
@@ -0,0 +1,56 @@
+
+        disposer.reset(new ComparatorDisposer(std::move(temporaryDisposer)));
+    }
+
```
There's a lot understand with this test... I'm only just learning about move semantics now, with the help of the new Josuttis book, and so it was really helpful to see this test. It took me a while to understand what was going on though.

As a suggestion, I think it could be simplified greatly by instead calling `ComparatorFactory::getComparatorForFileExtensionWithDot(".txt")` and checking that the return type is as expected... And removing the call to `Approvals::verify()` altogether.

I think the intention of the test would be clearer, and test-failures would be easier to understand...

Then the custom comparator for the test could be a trivial one, with no file-reading involved, or it could use one of the other provided comparators.

It would solve the other issues mentioned below, too...

claremacrae replied:
> It would solve the other issues mentioned below, too...

... mentioned above ...

claremacrae commented:
> I guess that the `"foo"`/`"bar"` comparisons in `TestComparator` could be confusing. Maybe it will be better if I changed this to case-insensitive comparison and rename `TestComparator` to `CaseInsensitiveComparator`?

Yes, that's a good point - for me, it did take a little while to work out the foo/bar thing... in the end, I did understand it - I've made a suggestion for an alternative approach in the code review...

p-podsiadly commented:
I've changed the test to use `FileApprover::verify`, the way it is done `FileApproverTest.cpp` - didn't think this earlier. The test should be more self-explanatory now.

claremacrae commented:
> I've changed the test to use `FileApprover::verify`, the way it is done `FileApproverTest.cpp` - didn't think this earlier. The test should be more self-explanatory now.

Yes, it's really clear now. Thanks!

Nice idea to not store the approved file in git. That should allow cloning in Windows fine now.

I wonder what will happen when the tests are run in Windows? I suspect it would exceed the file-name limit writing out the files and make the tests always fail.

claremacrae commented:
> I wonder what will happen when the tests are run in Windows? I suspect it would exceed the file-name limit writing out the files and make the tests always fail.

No - I'm wrong! Because you've supplied the names a.txt and b.txt, the test name is ignored and it's all good 🎉 !

claremacrae commented:
@p-podsiadly - this is great - I'm really happy to go ahead and merge it, if you are ready for me to do so?

I learned a lot from reading this code (in Catch and std C++) - thank you very much indeed!

p-podsiadly commented:
I think it's ready for merging. Glad I could help :)

claremacrae commented:
It's merged now - Thanks again!
